parser.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+)
+
+// Parser Structure: dictionary, sections
+// It's basically a map of a map, map of sections where each section maps to keys : values
+// map[section] --> returns a map where each key --> value
+type Parser struct {
+	dictionary map[string]map[string]string
+}
+
+// NewParser creates an instance of our Parser, which is iniitially an empty map of maps
+func NewParser() Parser {
+	return Parser{make(map[string]map[string]string)}
+}
+
+// ErrCouldNotOpen happens when file cannot be opened and provides file name
+var ErrCouldNotOpen error
+
+// ErrMissingValueAssignment happens when a key isn't followed by an = statement
+var ErrMissingValueAssignment error
+
+// ErrSectionNameMissingClosure happens when section name is missing the ] paranthesis
+var ErrSectionNameMissingClosure error
+
+// ErrWrongParanthesisOrder happens when section name starts by the wrong paranthesis ']'
+var ErrWrongParanthesisOrder error
+
+// ErrInvalidSectionName happens when section is written in a wrong form --> ex: sectionName]
+var ErrInvalidSectionName error
+
+// ErrCouldNotWriteToFile happens when file cannot be written to
+var ErrCouldNotWriteToFile error
+
+// LoadFromFile loads ini file
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromFile(fileName string) error {
+	var iniLines []string
+	var input io.Reader
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		ErrCouldNotOpen = fmt.Errorf("cannot open file: %s", fileName)
+		return ErrCouldNotOpen
+	}
+	defer file.Close()
+
+	input = file
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		iniLines = append(iniLines, scanner.Text())
+	}
+	return parsedMap.parserLogic(iniLines)
+}
+
+// LoadFromString loads ini script from a string
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromString(str string) error {
+	iniLines := strings.Split(str, "\n")
+	return parsedMap.parserLogic(iniLines)
+}
+
+func (parsedMap *Parser) parserLogic(iniLines []string) error {
+	var section, key, value string
+
+	for _, line := range iniLines {
+		if len(line) == 0 {
+			continue
+		} else if line[0] == '[' {
+			closingParaFound := false
+			for j, ch := range line {
```

Comment:
> you can check if the last char is ] instead of looping over the line

Done , since I edited the trimming line space at first so now it's valid (before it wasn't)